src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -30,43 +30,74 @@ namespace Microsoft.CodeAnalysis.CodeRefactorings;
     [ImportMany] IEnumerable<Lazy<CodeRefactoringProvider, CodeChangeProviderMetadata>> providers) : ICodeRefactoringService
 {
-    private readonly Lazy<ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>>> _lazyLanguageToProvidersMap = new Lazy<ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>>>(
-            () =>
-                ImmutableDictionary.CreateRange(
-                    DistributeLanguages(providers)
-                        .GroupBy(lz => lz.Metadata.Language)
-                        .Select(grp => KeyValuePairUtil.Create(
-                            grp.Key,
-                            new Lazy<ImmutableArray<CodeRefactoringProvider>>(() => [.. ExtensionOrderer.Order(grp).Select(lz => lz.Value)])))));
+    private readonly Lazy<ImmutableDictionary<ProviderKey, Lazy<ImmutableArray<CodeRefactoringProvider>>>> _lazyLanguageDocumentToProvidersMap =
+        new Lazy<ImmutableDictionary<ProviderKey, Lazy<ImmutableArray<CodeRefactoringProvider>>>>(() =>
+            ImmutableDictionary.CreateRange(
+                DistributeLanguagesAndDocuments(providers)
+                    .GroupBy(lz => new ProviderKey(lz.Metadata.Language, lz.Metadata.DocumentKind, lz.Metadata.DocumentExtension))
+                    .Select(grp => KeyValuePairUtil.Create(grp.Key,
+                        new Lazy<ImmutableArray<CodeRefactoringProvider>>(() => [.. ExtensionOrderer.Order(grp).Select(lz => lz.Value)])))));
+
     private readonly Lazy<ImmutableDictionary<CodeRefactoringProvider, CodeChangeProviderMetadata>> _lazyRefactoringToMetadataMap = new(() => providers.Where(provider => provider.IsValueCreated).ToImmutableDictionary(provider => provider.Value, provider => provider.Metadata));
 
     private ImmutableDictionary<CodeRefactoringProvider, FixAllProviderInfo?> _fixAllProviderMap = ImmutableDictionary<CodeRefactoringProvider, FixAllProviderInfo?>.Empty;
 
-    private static IEnumerable<Lazy<CodeRefactoringProvider, OrderableLanguageMetadata>> DistributeLanguages(IEnumerable<Lazy<CodeRefactoringProvider, CodeChangeProviderMetadata>> providers)
+    private static IEnumerable<Lazy<CodeRefactoringProvider, OrderableLanguageDocumentMetadata>> DistributeLanguagesAndDocuments(IEnumerable<Lazy<CodeRefactoringProvider, CodeChangeProviderMetadata>> providers)
     {
         foreach (var provider in providers)
         {
             foreach (var language in provider.Metadata.Languages)
             {
-                var orderable = new OrderableLanguageMetadata(
-                    provider.Metadata.Name, language, provider.Metadata.AfterTyped, provider.Metadata.BeforeTyped);
-                yield return new Lazy<CodeRefactoringProvider, OrderableLanguageMetadata>(() => provider.Value, orderable);
+                foreach (var documentKind in provider.Metadata.DocumentKinds)
+                {
+                    if (Enum.TryParse<TextDocumentKind>(documentKind, out var kind))
```

Comment:
We probably should handle a failure here in some way.  Potentially the enum parsing should move to the creation of the metadata object itself?  I believe that would cause a MEF error if it fails.

Otherwise this should possibly just throw if we cannot parse the kind.

Reply:
Metadata is being passed thru MEF pipeline and converted to text. BTW ExportCodeRefactoringProviderAttribute will throw if it kinds are not enum values.

Reply:
Could this then be a `Enum.Parse` then?